### PR TITLE
Clone instances of `User` in `Msg` to avoid unintentional mutations

### LINKED ISSUE #1765: Nick and modes for users when a message is sent, isn't kept
Everything looks fine if lounge is open while receiving messages, but after a reload, messages display the user's *current* mode and nick, even if they didn't have that nick or mode when the message was sent.

Running master (c9e340e)

before reload:

after reload:

### PR DESCRIPTION
**Pull request by astorije**
Fixes #1765.

This supersedes #1767. I find this way preferable because we have better ways to ensure this security. Also, test case!

Ah, if only we were using https://github.com/facebook/immutable-js/ :trollface:

### CODE REVIEW

File: src/models/msg.js [hunk 1]
```diff
@@ -6,8 +6,17 @@ var id = 0;
...
+			["from", "invited", "target"].forEach((prop) => {
+				if (attr[prop]) {
+					this[prop] = _.cloneDeep(attr[prop]);
+				}
+			});
+		}
+
 		_.defaults(this, attr, {
-			from: "",
+			from: {},
```
Comment (astorije) [inline src/models/msg.js]: This should fix some other edge cases we missed IMO.

**Comment by astorije** [comment]
@PolarizedIons, could you test this PR against the bug you reported in #1765?

File: src/models/msg.js [hunk 2]
```diff
@@ -6,6 +6,15 @@ var id = 0;
 
 class Msg {
 	constructor(attr) {
+		// Some properties need to be copied in the Msg object instead of referenced
+		if (attr) {
+			["from", "invited", "target"].forEach((prop) => {
```
Comment (xPaw) [inline src/models/msg.js]: If we're going this route, change invited to target in invite messages so that it's somewhat consistent.

File: src/models/msg.js [hunk 3]
```diff
@@ -6,6 +6,15 @@ var id = 0;
...
+				if (attr[prop]) {
+					this[prop] = _.cloneDeep(attr[prop]);
+				}
```
Comment (xPaw) [inline src/models/msg.js]: As with previous PR, store only nick and mode because we don't want to copy all the other unnecessary stuff (away message etc)

**Comment by PolarizedIons** [comment]
I can confirm this also fixes the issue. I have no problems that if want to take over the issue :) I guess I'll close mine in the meantime.

File: src/models/msg.js (hunk 2 above)
Comment (astorije) [reply src/models/msg.js]: Gotcha, will do.

File: src/models/msg.js (hunk 3 above)
Comment (astorije) [reply src/models/msg.js]: Fair enough. I figured it would be better for future sake to store the whole object, but I guess we don't have to at the moment.

File: src/models/msg.js (hunk 2 above)
Comment (astorije) [reply src/models/msg.js]: Done! Would you mind testing a bit to make sure I didn't break invite stuff? I can't seem to be able to display "xxx invited yyy", only "xxx invited you".

File: src/models/msg.js (hunk 3 above)
Comment (astorije) [reply src/models/msg.js]: Done.
